src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -567,10 +579,15 @@ void TemporalMemory::activateDendrites(bool learn,
     NTA_CHECK( extraWinners.size() != 1 || extraWinners[0] != SENTINEL )
         << "TM.ActivateDendrites() missing argument extraWinners!";
 
+    //add extra active
+    auto& activeVec = activeCells_.getSparse();
     for(const auto &active : extraActive) {
       NTA_ASSERT( active < extra_ );
-      activeCells_.push_back( static_cast<CellIdx>(active + numberOfCells()) ); 
+      activeVec.push_back( static_cast<CellIdx>(active + numberOfCells()) ); 
```

Comment:
Yes, make two new vectors named something like presynActive & presynWinner.  Copy the TM's active and winner cells into it, and concatenate the extra inputs into those vectors too.

Reply:
> The external predictive inputs are not part of the active cells of this TM, and shouldn't be added to the activeCells. This is a hack, really it should concatenate both the activeCells and the externalPredictiveCells into a new vector, instead of repurposing an existing vector. 

This is the culpit! 

> The external predictive inputs are not part of the active cells 

ok, let's do it properly. and not add them to activeCells_. 
Proposal: Should we use a member SDR extraActivations_ ? 
 - would allow to (again) unite the activateDendrites API (no version for extra*) 
 - users would call setExtra*Cells(SDR) to set them. After compute, the SDR gets cleared again. 

> are not part of the active cells of this TM

but they do refer to cells in this TM, right? 
Sorry, again, what is the meaing of externalActive/WinnerCells? I thought it's an "deus ex" activation of the TM's cells. 

So, if I have a TM with only 3 cells {0,1,2}. we now encode extras with offset, so {0,1,2, ex3, ex4, ex5}. 
Assume this state (dense repr): {101 |extra: 111}
Does this translate to: 
 - {111} active cells?
 - or only +3 number to pass threshold, so 5 active cells (2 from this TM). 

Another question, how should we treat the extra activations in context of anomalies? 
- include in the computation,
- or ignore?

Reply:
> but they do refer to cells in this TM, right?
Sorry, again, what is the meaing of externalActive/WinnerCells? I thought it's an "deus ex" activation of the TM's cells.

No, external here means that it comes from outside of yhe TM.

  For example i could have 2 brain areas with 2 TMs which are looking different sensors, and one TM can help the othermake predictions.  

Another Example is to model L4 with a TM and to generate a location signal and feed it to the TM.  Numenta did this in their "columns" paper.

Reply:
> Another question, how should we treat the extra activations in context of anomalies?

They should ve included if they are given.

Reply:
@ctrl-z-9000-times would you mind finishing the extra* issue, please? I'm still not completely getting the picture, so I'd make mistakes in the implementation. 

Some notes: 
- please rename extra* to external* (or presynaptic*)
- nit, could be extended to N external sources with vector<SDR> 
- could you improve doc on winnerCells, please? 
  - should TMReg default output actually be winner cells? 
  - similar, Anomaly, should be computed on winner-prev predicted? (instead of active-prev predicted?) 
  - why is there extra Winner/Active, but not Predictive? 
- consider having TM.setExtraWinner/Active(SDR) instead of the (optional) args to activateDendrites() ?

Reply:
Another view is, that it's valid and working to have activeCells_,winnerCells_ as vector due to the extra* functionality, and we cannot do that with SDR that strictly checks the members are within its bounds. Also, having active/winners as SDR does not give some big advantage. So we could wery well just close this PR. 

thoughts?

Reply:
I can work on this a bit.
I will change the name from extra to something more descriptive.

> TMReg default output

The TM has multiple outputs, all of which are valid.  IMO its better to not have a default at all and to force the user to specify what they want

Reply:
:+1: thank you!

> MO its better to not have a default at all and to force the user to specify what they want 

you're talking about TM, I agree. I was talking about TMRegion where NetworkAPI forces us to take a default output. Is it that winner cells would be generally more useful/"the nicest"?

Reply:
>  Is it that winner cells would be generally more useful?

It depends on what you're doing. 

Winner cells are the cells which the TM has designated to represent the current state of its world, so they should be used for learning.

Reply:
> [winner cells]  represent the current state of its world, 

Sounds that should be the default output of TMRegion, I'll update that. 
Should anomaly also be made from Winners/Predictive?